dipy/denoise/noise_estimate.py

```diff
@@ -61,6 +67,106 @@ def piesno(data, N=1, alpha=0.01, l=100, itermax=100, eps=1e-5, return_mask=Fals
     background and 2. The data is a repetition of the same measurements
     along the last axis, i.e. dMRI or fMRI data, not structural data like T1/T2.
 
+    This function processes the data slice by slice, but makes a global
+    estimation of the noise. Use piesno_3D to get a slice by slice estimation
+    of the noise, as in spinal cord imaging for example.
+
+    References
+    ------------
+
+    .. [1] Koay CG, Ozarslan E and Pierpaoli C.
+    "Probabilistic Identification and Estimation of Noise (PIESNO):
+    A self-consistent approach and its applications in MRI."
+    Journal of Magnetic Resonance 2009; 199: 94-103.
+
+    .. [2] Koay CG, Ozarslan E and Basser PJ.
+    "A signal transformational framework for breaking the noise floor
+    and its applications in MRI."
+    Journal of Magnetic Resonance 2009; 197: 108-119.
+    """
+
+    # This method works on a 2D array with repetitions as the third dimension,
+    # so process the dataset slice by slice.
+
+    if data.ndim < 3:
+        raise ValueError("This function only works on datasets of at least 3 dimensions.")
+
+    if data.ndim == 4:
+
+        sigma = np.zeros(data.shape[-2], dtype=np.float32)
+        mask_noise = np.zeros(data.shape[:-1], dtype=np.bool)
+
+        for idx in range(data.shape[-2]):
+            sigma[idx], mask_noise[..., idx] = piesno_3D(data[..., idx, :], N,
+                                                          alpha=alpha, l=l, itermax=itermax, eps=eps)
+
+        # Take the mode of all the sigmas from each slice as the best estimate,
+        # this should be stable with more or less 50% of the guesses at the same value.
+        sigma, num = mode(sigma, axis=None)
```

Comment:
Note to myself - change this to return the full array, as it's just plain misleading to return a single value on real data.